tests/inputs/stmtTests.mg

```diff
@@ -17,4 +17,107 @@ implementation StmtTests = {
         // Invalid assignment
         t1 = mkT2();
     }
-};
+}
+
+// TODO: add more tests
+implementation VarExprTests = {
+
+    require type T;
+    require type U;
+
+    procedure conflicting_type_annotation() = {
+        var x: T;
+        // should succeed
+        x: T;
+        // should fail
+        x: U;
+    }
+}
+
+
+// TODO: add more tests
+implementation CallExprTests = {
+
+    require type T;
+    require type U;
+
+    function overload_return_type(): T;
+    function overload_return_type(): U;
+
+    procedure call_functions() = {
+
+        // should succeed
+        var explicit_T: T = overload_return_type();
+
+        // should succeed
+        var explicit_U: U = overload_return_type();
+
+        // should fail
+        var implicit_type = overload_return_type();
+    }
+}
+implementation LetExprTests = {
+
+    require type T;
+    require type U;
+
+    procedure var_assignment() = {
+        // should succeed
+        var x: T;
+        // should fail
+        var x: U;
+
+    }
+
+    procedure var_declaration_infer_type() = {
+        // should succeed
+        // TODO(mariuslarnoy 2021/07/19): this is not implemented yet
+        var x;
+    }
```

Comment:
I don't think this should succeed. If we had a more sophisticated type checking pass, a body such as
```
var x; x = some_fun_of_type_T();
```
could succeed.

But I am not sure if we even want that anymore. Your specific example here would never succeed, because a type for `x` can never be deduced.

Reply:
Yes the specific example would never succeed, bad (or rather lazy) pick by me. Updated it now.